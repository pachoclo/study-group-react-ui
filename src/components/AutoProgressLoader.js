import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'

const defaults = {
  initialProgress: 5,
  intervalMillis: 500,
  increment: 5
}

/**
 * Renders a loader that automatically increments its progress
 *
 * @param {number} intervalMillis specifies how often the progress should be increased
 */
export const AutoProgressLoader = ({ intervalMillis = defaults.intervalMillis }) => {
  const [progress, setProgress] = useState(defaults.initialProgress)

  useEffect(() => {
    const intervalId = setInterval(() => {
      setProgress(currentProgress => {
        const newProgress = currentProgress + 5
        return newProgress <= 100 ? newProgress : currentProgress
      })
    }, intervalMillis)

    return function cleanUp () {
      clearInterval(intervalId)
      setProgress(100)
    }
  }, [intervalMillis])

  return (
    <div className='progress' style={{ height: '4px', backgroundColor: 'gray', borderRadius: 0 }}>
      <div
        className='progress-bar progress-bar-striped progress-bar-animated bg-warning'
        role='progressbar'
        aria-valuenow={progress}
        aria-valuemin='0'
        aria-valuemax='100'
        style={{ width: `${progress}%` }}
      />
    </div>
  )
}

AutoProgressLoader.propTypes = {
  intervalMillis: PropTypes.number
}

export default AutoProgressLoader
