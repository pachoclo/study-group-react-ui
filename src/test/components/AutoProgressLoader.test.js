import React from 'react'
import AutoProgressLoader from '../../components/AutoProgressLoader'
import { shallow, mount } from 'enzyme'
import { act } from 'react-dom/test-utils'

describe('AutoProgressLoader Test', () => {
  let container

  beforeEach(() => {
    container = document.createElement('div')
    document.body.appendChild(container)
  })

  afterEach(() => {
    document.body.removeChild(container)
    container = null
  })

  it('should trigger setInterval and clearInterval', () => {
    jest.useFakeTimers()
    let component
    act(() => {
      component = mount(<AutoProgressLoader intervalMillis={5} />)
    })
    expect(component).toMatchSnapshot()
    expect(setInterval).toHaveBeenCalled()

    act(() => {
      jest.advanceTimersByTime(3000)
    })

    act(() => {
      component.unmount()
    })
    expect(clearInterval).toHaveBeenCalled()
  })
})
