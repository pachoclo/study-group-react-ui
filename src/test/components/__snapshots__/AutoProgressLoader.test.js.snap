// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoProgressLoader Test should trigger setInterval and clearInterval 1`] = `
<AutoProgressLoader
  intervalMillis={5}
>
  <div
    className="progress"
    style={
      Object {
        "backgroundColor": "gray",
        "borderRadius": 0,
        "height": "4px",
      }
    }
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow={5}
      className="progress-bar progress-bar-striped progress-bar-animated bg-warning"
      role="progressbar"
      style={
        Object {
          "width": "5%",
        }
      }
    />
  </div>
</AutoProgressLoader>
`;
